### PR TITLE
[QUIC] Call silent shutdown in case `CloseAsync` failed.

### LINKED ISSUE #78641: `QuicConnection.DisposeAsync` throws `OperationCanceledException`
### Description

The attached test case demonstrates that `QuicConnection.DisposeAsync` throws `OperationCanceledException`. 

Also, looking at the implementation of `QuicConnection.DisposeAsync`, `DisposeAsync` is not a silent close as described by the documentation. It's a graceful close that sends a close frame with the `DefaultCloseErrorCode`. 

This also raises the question of what `DisposeAsync` should do if `CloseAsync` is canceled.

### Reproduction Steps

Run the attach program with a Kestrel server.

[quic-client.zip](https://github.com/dotnet/runtime/files/10057379/quic-client.zip)

### Expected behavior

```
operation canceled
```

### Actual behavior

```
operation canceled
Unhandled exception. System.OperationCanceledException: The operation was canceled.
   at System.Net.Quic.ValueTaskSource.TryComplete(Exception exception)
   at System.Net.Quic.ValueTaskSource.<>c.<TryInitialize>b__10_0(Object obj, CancellationToken cancellationToken)
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool)
   at System.Threading.TimerQueue.FireNextTimers()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
--- End of stack trace from previous location ---
   at System.Net.Quic.ValueTaskSource.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Quic.QuicConnection.DisposeAsync()
   at Program.Main(String[] args) in /home/vagrant/workspace/quic-client/Program.cs:line 35
   at Program.Main(String[] args) in /home/vagrant/workspace/quic-client/Program.cs:line 35
   at Program.<Main>(String[] args)
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by ManickaP**
1. Makes cancellation in `ValueTaskSource` temporary, which enables receiving the expected MsQuic signal even after `cancellationToken` fired. In this case `SHUTDOWN_COMPLETE`.
2. If `ValueTaskSource` is cancelled, remember what happened to it before the caller read the OCE so that it can be restored to the task afterwards.
3. When the OCE is delivered, restore `ValueTaskSource` to the remembered state (awaiting or completed with the result).
4. Lastly, examine the state of `_shutdownTcs` in `Dispose` and decide whether the silent shutdown needs to be called.

Changes are in ValueTaskSource and QuicConnection. Other changes are cosmetic, not functional (comments, code formatting).

Fixes https://github.com/dotnet/runtime/issues/78641

I also did a manual testing with cancellation between CloseAsync --> connection.Shutdown and SHUTDOWN_COMPLETED by inserting delay to the event handler, but this as this is timing sensitive it's impossible to do reliably in a test.

### CODE REVIEW

**Comment by rzikm** [comment]
I can still get DisposeAsync to throw like this.

```cs
var (clientOptions, _, listenerOptions) = RuntimeSandbox.Helpers.GetDefaultOptions();

await using var pair = await RuntimeSandbox.Helpers.GetConnectedPair(clientOptions, listenerOptions);
var (client, server) = pair;

var cts = new CancellationTokenSource();
var task = client.CloseAsync(0).AsTask();

cts.Cancel(); // throws even without this line
await client.DisposeAsync();
// Unhandled exception. System.InvalidOperationException: Operation is not valid due to the current state of the object.
//    at System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore`1.OnCompleted(Action`1 continuation, Object state, Int16 token, ValueTaskSourceOnCompletedFlags flags)
//    at System.Net.Quic.ValueTaskSource.System.Threading.Tasks.Sources.IValueTaskSource.OnCompleted(Action`1 continuation, Object state, Int16 token, ValueTaskSourceOnCompletedFlags flags)
//    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AwaitUnsafeOnCompleted[TAwaiter](TAwaiter& awaiter, IAsyncStateMachineBox box)
// --- End of stack trace from previous location ---
//    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
//    at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
//    at System.Threading.ThreadPoolWorkQueue.Dispatch()
//    at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
//    at System.Threading.Thread.StartCallback()
```

Do we care? If not then LGTM.

**Comment by CarnaViire** [comment]
@rzikm it seems like it's a different case, not related to this cancellation issue? (do I understand correctly that this "Operation is not valid due to the current state of the object" happens in main, and _continues_ to happen after the PR)?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs [hunk 1]
```diff
@@ -173,5 +187,34 @@ void IValueTaskSource.OnCompleted(Action<object?> continuation, object? state, s
...
+                    _valueTaskSource.Reset();
+                    if (state == State.Completed)
+                    {
+                        if (_exception is not null)
+                        {
+                            _valueTaskSource.SetException(_exception);
+                        }
+                        else
+                        {
+                            _valueTaskSource.SetResult(true);
```
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs:213]: this all is soooo confusing now 😖😵‍💫😅

is this ValueTaskSource only ever used for closing the connection? or is it used for streams as well?

Reading/writing I believe use ResettableValueTaskSource which is _completely_ unrelated to this one, right?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs [hunk 2]
```diff
@@ -173,5 +187,34 @@ void IValueTaskSource.OnCompleted(Action<object?> continuation, object? state, s
...
+        finally
+        {
+            lock (this)
+            {
+                State state = _state;
+
+                // In case of a cancellation, reset the task and set the stored results if necessary.
+                if (_valueTaskSource.GetStatus(_valueTaskSource.Version) == ValueTaskSourceStatus.Canceled)
+                {
+                    _valueTaskSource.Reset();
```
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs:204]: How and when we can consume the _actual_ saved exception after that? 🤔

**Comment by ManickaP** [comment]
> I can still get DisposeAsync to throw like this.

Just FYI, this happens without this change as well. I'll try looking into it if it's something small that I could put into this PR and if not I'll file separate issue.

**Comment by rzikm** [comment]
> But I think we _should_ care 🤔 I don't think it's ok to throw if dispose and/or close are called in parallel.

I think we should never throw from Dispose(Async)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs (hunk 1 above)
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs]: Changes in VTS reverted.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs (hunk 2 above)
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs]: Changes in VTS reverted.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs [hunk 3]
```diff
@@ -467,7 +467,7 @@ public ValueTask CloseAsync(long errorCode, CancellationToken cancellationToken
     {
         ObjectDisposedException.ThrowIf(_disposed == 1, this);
 
-        if (_shutdownTcs.TryInitialize(out ValueTask valueTask, this, cancellationToken))
+        if (Interlocked.CompareExchange(ref _shutdownTcs, new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously), null) == null)
```
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:470]: You are creating an instance here every time regardless whether the exchange happened or not... is that expected?

Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:470]: I did not bother with optimizing it. It'll be at most 2 instances (in Gen0) in vain per connection lifetime. I can change it to use another field as a sentinel  (e.g. `int shutdownCalled`) or to preallocate this in other field for a swap. Or just not guard against multiple `Shutdown` calls :shrug:

**Review by CarnaViire — APPROVED** [review]
LGTM

**Comment by ManickaP** [comment]
Just FYI, using ordinary TCS brought back an issue with GC eating non-rooted `QuicConnection` while waiting on `SHUTDOWN_COMPLETE`, so I'll be re-thinking this once again. I'll ping you for re-review when the fix is up.

**Review by rzikm — APPROVED** [review]

**Comment by ManickaP** [comment]
Failure is unrelated and known.

**Pull request merged**